instant-segment/examples/merge.rs

```diff
@@ -0,0 +1,246 @@
+use std::io::{BufRead, BufReader, BufWriter};
+use std::str::FromStr;
+
+use ahash::{AHashMap as HashMap, AHashSet as HashSet};
+use rayon::iter::{IntoParallelIterator, ParallelIterator};
+use smartstring::alias::String as SmartString;
+
+fn main() {
+    let word_list = read_word_list();
+    process_unigrams(&word_list);
+    process_bigrams(&word_list);
+}
+
+/// Read bigrams from the input file parts, filter them, and write to file
+fn process_bigrams(word_list: &HashSet<SmartString>) {
+    let bigrams = (0..BIGRAM_PARTS)
+        .into_par_iter()
+        .map(|part| {
+            let fname = format!("data/cache/eng-2-{:05}-{:05}.txt", part, BIGRAM_PARTS);
+            let f = File::open(&fname).unwrap();
+            let mut reader = BufReader::with_capacity(4 * 1024 * 1024, f);
+
+            let mut ln = String::new();
+            let mut bigrams = HashMap::new();
+            loop {
+                ln.clear();
+                match reader.read_line(&mut ln) {
+                    Ok(0) => break,
+                    Err(e) => {
+                        eprintln!("error: {:?}", e);
+                        break;
+                    }
+                    _ => {}
+                }
+
+                let mut iter = ln.trim().split('\t');
+                let words = match iter.next() {
+                    Some(word) => word,
+                    None => continue,
+                };
+
+                let mut word_iter = words.split(' ');
+                let word1 = match word_iter.next() {
+                    Some(word) => word,
+                    _ => continue,
+                };
+
+                let word1 = match normalize(word1, word_list) {
+                    Some(word) => word,
+                    _ => continue,
+                };
+
+                let word2 = match word_iter.next() {
+                    Some(word) if word_list.contains(word) => word,
+                    _ => continue,
+                };
+
+                let word2 = match normalize(word2, word_list) {
+                    Some(word) => word,
+                    _ => continue,
+                };
+
+                let mut matches = 0;
+                for year_data in iter {
+                    let mut parts = year_data.split(',');
+                    if parts.next().unwrap() < START_YEAR {
+                        continue;
+                    }
+                    matches += usize::from_str(parts.next().unwrap()).unwrap();
+                }
+
+                if bigrams.capacity() == 0 {
+                    // While it's not uncommon for a part to result in 0 words, the average for
+                    // parts that contain more than 0 is about 300k, median is about 350k. Allocate
+                    // a decent chunk immediately to avoid too many intermediate reallocations.
+                    bigrams.reserve(256 * 1024)
+                }
+
+                *bigrams.entry((word1, word2)).or_default() += matches;
```

Comment:
As a Rust newbie, this immediately obvious what this is doing, can you explain?

Reply:
I'm using this: https://doc.rust-lang.org/std/collections/hash_map/enum.Entry.html#method.or_default.

It's kind of like Python's `dict.setdefault()`. `entry()` returns an entry which can be `Vacant` or `Occupied`. `or_default()` will return a `&mut` to the value, inserting the `Default::default()` for the value's type if the entry is currently `Vacant`. `Default` is implemented for `usize` and yields 0.

Reply:
Cool, thanks